kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/SubscriberServiceImpl.java

```diff
@@ -67,6 +125,113 @@ public void update(Subscriber subscriber) {
         }
     }
 
+    @Override
+    public void updateMsisdnForSubscriber(Subscriber subscriber, MctsBeneficiary beneficiary, Long newMsisdn) {
+        SubscriptionPackType packType;
+        packType = (beneficiary instanceof MctsChild) ? SubscriptionPackType.CHILD : SubscriptionPackType.PREGNANCY;
+
+        Subscriber subscriberWithMsisdn = subscriberDataService.findByCallingNumber(newMsisdn);
+        if (subscriberWithMsisdn != null) {
+            // this number is in use
+            if (subscriptionService.getActiveSubscription(subscriberWithMsisdn, packType) != null) {
+                // in fact, it's in use for this pack -- reject the subscription
+                throw new IllegalStateException(
+                        String.format("Can't change subscriber's MSISDN (%s) because it is already in use for %s pack",
+                                newMsisdn, packType));
+            }
+        }
+
+        // do the update -- by creating a new Subscriber object and re-linking the beneficiary and subscription to it
+
+        Subscriber newSubscriber = new Subscriber(newMsisdn, subscriber.getLanguage(), subscriber.getCircle());
+        Subscription subscription = subscriptionService.getActiveSubscription(subscriber, packType);
+        subscriber.getSubscriptions().remove(subscription);
+        newSubscriber.getSubscriptions().add(subscription);
+        subscription.setSubscriber(newSubscriber);
+
+        if (packType == SubscriptionPackType.CHILD) {
+            subscriber.setChild(null);
+            newSubscriber.setChild((MctsChild) beneficiary);
+            newSubscriber.setDateOfBirth(subscriber.getDateOfBirth());
+        } else {
+            subscriber.setMother(null);
+            newSubscriber.setMother((MctsMother) beneficiary);
+            newSubscriber.setLastMenstrualPeriod(subscriber.getLastMenstrualPeriod());
+        }
+
+        subscriberDataService.create(newSubscriber);
+        subscriberDataService.update(subscriber);
+        subscriptionDataService.update(subscription);
+    }
+
+    @Override
+    public Subscription updateOrCreateMctsSubscriber(MctsBeneficiary beneficiary, Long msisdn, DateTime referenceDate,
+                                                     SubscriptionPackType packType) {
+        District district = beneficiary.getDistrict();
+        Language language = (Language) districtDataService.getDetachedField(district, "language");
+        Subscriber subscriber = getSubscriber(msisdn);
+
+        SubscriptionPack pack = subscriptionPackDataService.byType(packType);
+
+        // TODO: Handle the case in which the MCTS beneficiary already exists but with a different phone number
```

Comment:
Should we have a ticket number for that?
Are we waiting for input before we can proceed?